Refactor file permissions on rm tree on Windows

johnwparent opened the pull request:
Resolves #29713

johnwparent committed:
Refactor file permissions on rm tree on Windows

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1117,6 +1117,18 @@ def visit_directory_tree(root, visitor, rel_path='', depth=0):
 
 
+@system_path_filter
+def lightweight_traverse_tree(root, visit=None, *args, **kwargs):
```
This function could benefit from fully implementing the visitor for `visit_directory_tree`:

* Avoid entering symlinks because we only need to visit each file once
* Do we need to update permissions on the symlinks themselves? (that isn't required on Linux)

Also regarding the name: the term `lightweight` is distracting and doesn't provide much info (e.g. one might think that this is the "lightweight" version of `traverse_tree` which it is not, and furthermore understanding what lightweight means requires reading both functions). I think `visit_recursive` would be more clear (and the docstring could say "this is like visit_directory_tree, but follows all symlinks).

johnwparent replied:
I'll have to investigate regarding the symlinks themselves, but I think if we're going to be implementing the Visitor, we may as well scrap this method (I thought we may want to do that), unless I'm misunderstanding you? 

Fully agreed on the name, I wasn't sure if this method would be dropped in preference of the actual traverse tree method.

scheibelp replied:
> we're going to be implementing the Visitor, we may as well scrap this method (I thought we may want to do that), unless I'm misunderstanding you?

To confirm: implementing the visitor seems like the best way to proceed IMO. I figure it still might make sense to have a separate method, unless `os.walk` can for example achieve the basic behavior of "visit each file in the tree, don't follow symlinks".

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1190,20 +1199,15 @@ def remove_linked_tree(path):
+    # Ensure proper permissions
+    if is_windows:
+        def make_removable(path):
+            os.chmod(path, stat.S_IWUSR)
```
On linux the typical approach for modifying file permissions is

```
old_permissions = os.stat(the_file)
os.chmod(the_file, old_permissions | new_permission)
```

Does that not work on Windows?

johnwparent replied:
From the docs: 

> Note Although Windows supports [chmod()](https://docs.python.org/3/library/os.html#os.chmod), you can only set the file’s read-only flag with it (via the stat.S_IWRITE and stat.S_IREAD constants or a corresponding integer value). All other bits are ignored.

scheibelp reviewed (changes requested):
I have a couple of requests

haampie commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1178,26 @@ def remove_if_dead_link(path):
+    def before_visit_symlinked_dir(self, root, rel_path, depth):
+        return False
+
+    def after_visit_symlink(self, root, rel_path, depth):
+        self.caller(os.path.join(root, rel_path))
```
I think this is a typo? Maybe we should just implement a dummy interface class for the visitor, so that you don't have to implement `after_visit_symlinked_dir` when you don't need it.

johnwparent replied:
Yes.
I agree, I was surprised when I didn''t find one already, I'll create a `GenericVisitor` class to encompass that

haampie commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1190,22 +1208,18 @@ def remove_linked_tree(path):
+            def make_removable(path):
+                os.chmod(path, stat.S_IWUSR)
+
+            visit_directory_tree(path, CallingVisitor(make_removable))
```
Note that you're still calling chmod on symlinked files possibly outside of `path`, as well as on dangling symlinks, which may fail.

haampie replied:
Maybe it's just easiest to extend the visitor interface with `visit_symlinked_file` to force the user of visit_directory_tree to think about handling symlinks.

haampie commented on lib/spack/llnl/util/link_tree.py:
```diff
@@ -44,7 +44,7 @@ def __init__(self, dst, src_a=None, src_b=None):
 
 
-class SourceMergeVisitor(object):
+class SourceMergeVisitor(GenericVisitor):
```
`visit_symlinked_file` should call `visit_file` here

johnwparent replied:
I'll make that update. Thanks.

haampie replied:
if there are no failing tests that'd be worrying ;p

haampie commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1179,46 @@ def remove_if_dead_link(path):
 
 
+class GenericVisitor(object):
+    def __init__(self, callable):
+        pass
```
```suggestion
```

scheibelp replied:
I agree with this suggestion (I don't think the `__init__` in the interface is useful for enforcing policy in subclasses)

johnwparent committed:
RegisterVisitor call symlinked_file

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1097,16 +1097,26 @@ def visit_directory_tree(root, visitor, rel_path='', depth=0):
-            islink, isdir = f.is_symlink(), f.is_dir()
+            islink = f.is_symlink()
+            # Windows issue with reading symlinked archive
+            # results in [WinError 1921]
```
(question) Does `isdir` always fail on symlinks? Or when you say archive you mean for a symlink to a file that is for example a `.zip`?

johnwparent replied:
`isdir` will fail on symlinks to a file that is a `.zip` or something similar.

haampie replied:
Isn't that a Python bug?

johnwparent replied:
On further investigation, this may be due to how broken or self referencing symlinks are handled on Windows, rather than a bug.

scheibelp replied:
We spoke about updating the comment here to provide more context on when things fail on Windows

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1097,16 +1097,26 @@ def visit_directory_tree(root, visitor, rel_path='', depth=0):
 
-        if not isdir:
-            # Handle files
+        # handle symlinked files
```
I think this comment needs to be edited: this first `if` is specifically for files that are *not* symlinks.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1187,69 @@ def remove_if_dead_link(path):
+        pass
+
+    def before_visit_dir(self, root, rel_path, depth):
+        pass
```
It looks like this is an interface, so:

* The name `Generic` is misleading: `VisitorInterface` would be more useful
* These functions should raise `NotImplementedError`
* The interface should have a docstring for the class and for each function explaining the intent

johnwparent replied:
The intent was to create a stub class rather than a true interface as per [this comment](https://github.com/spack/spack/pull/29714#discussion_r838676576), however if you feel we should be implementing a true interface class here that's fine with me.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1187,69 @@ def remove_if_dead_link(path):
+        pass
+
+
+class CallingVisitor(GenericVisitor):
```
Firstly, I see this used in one spot: `remove_dead_links`. I don't think you actually have to traverse all symlink dirs in order to find dead links, so I'm not sure if this implementation is needed.

Secondly (and this part is only important if you disagree with the first part and think this class should remain), IMO this name doesn't explain well what this implementation does: all visitors call their supplied callback on traversed files/dirs. Probably more useful is to set the name (and docstring) to explain that this traverses all files, directories, and symlinks.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1187,69 @@ def remove_if_dead_link(path):
+        self.caller(os.path.join(root, rel_path))
+
+
+class MakeWinWriteableVisitor(GenericVisitor):
```
(naming) While this visitor is only used for this purpose, it probably makes sense to name this more based on it's behavior WRT how it decides to visit files/dirs/symlinks.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1187,69 @@ def remove_if_dead_link(path):
+    def visit_file(self, root, rel_path, depth):
+        pass
+
+    def visit_symlinked_file(self, root, rel_path, depth):
```
One thing to clarify in the docstring for this function: will it invoke the callback on the symlink path, or on the path of the target of the symlink?

scheibelp reviewed (changes requested):
I have some requests related to the overall organization of the interface and the behavior related to entering symlink directories (I suspect that generally isn't necessary).

johnwparent committed:
Refactor to shutil callback

Extend to directory layout

Update docstring

Docstring should reflect that method will raise
on non Windows platforms

Update doc for force delete

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1187,16 @@ def remove_if_dead_link(path):
         os.unlink(path)
 
 
+def errorRemoveReadonly(func, path, exc):
```
(minor) can you rename this to follow the general scheme used elsewhere (all-lowercase with `_` between words)?

johnwparent replied:
Adressed

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1190,20 +1207,8 @@ def remove_linked_tree(path):
-    kwargs = {'ignore_errors': True}
-    if is_windows:
-        kwargs = {'onerror': onerror}
+    kwargs = {'ignore_errors': False,
```
Why does this undo https://github.com/spack/spack/commit/048a0de35bd791f233d0a6f676d841d85e75aae7 (part of that was to explicitly restore `ignore_errors=True` on Linux)?

I think this could work because of https://github.com/spack/spack/pull/29690/files#r833667276 (i.e. the updates to `Stage.purge`). But that PR still restored `ignore_errors=True`. @alalazo is there a reason you did both of those things? Was it to support calling `remove_linked_tree` in other contexts?

johnwparent replied:
I believe this was addressed from my end in our last meeting.

scheibelp replied:
I realized in my latest review that this doesn't undo https://github.com/spack/spack/commit/048a0de35bd791f233d0a6f676d841d85e75aae7 so I'm going to mark this as resolved (that review brought up a related concern but that has its own thread)

scheibelp reviewed (changes requested):
Minor naming request and also I wanted to check about the `ignore_errors` edits in relation to https://github.com/spack/spack/pull/29690.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1187,16 @@ def remove_if_dead_link(path):
+        # change the file to be readable,writable,executable: 0777
+        os.chmod(path, stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO)
+        # retry
+        func(path)
```
I had to stare at this for a while to realize that this effectively preserves `ignore_errors=True` on Linux because this function doesn't raise any exceptions (or do anything) on Linux. I think it would be more straightforward to update how the `kwargs` are built in `remove_linked_tree` (i.e. on Linux `ignore_errors=True` and on Windows `ignore_errors=False` + `onerror=...`).

scheibelp reviewed (changes requested):
I have a couple more requests

scheibelp commented on lib/spack/spack/directory_layout.py:
```diff
@@ -360,7 +360,8 @@ def remove_install_directory(self, spec, deprecated=False):
             try:
-                shutil.rmtree(path)
+                shutil.rmtree(path, ignore_errors=False,
+                              onerror=fs.error_remove_readonly)
```
This creates implied `ignore_errors=True` on non-Windows systems so:

* Update `error_remove_readonly` so it throws an error if it is ever called on a non-Windows system (to avoid similar issues sneaking by in the future)
* You'll need to do the same `kwargs` manipulation here as was done in `remove_linked_tree`

johnwparent replied:
I keep missing this change, thank you, my apologies. These changes have been made.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1097,15 +1097,26 @@ def visit_directory_tree(root, visitor, rel_path='', depth=0):
+            try:
+                isdir = f.is_dir()
+            except OSError:
+                isdir = False
```
Thanks for updating the comment! I'm worried this is too broad of a catch: `f.is_symlink` is called first, so would probably trigger most errors (e.g. permissions errors) befor `.is_dir` could, but I still think it might be worth taking care to check the error code (if possible) and/or `raise`ing on non-Windows systems. Does that seem reasonable?

johnwparent replied:
raise-ing on non Windows makes sense, and I'll dig into whether or not we can get more specificity on the Windows side.

scheibelp commented on lib/spack/spack/test/llnl/util/link_tree.py:
```diff
@@ -14,8 +13,8 @@
-pytestmark = pytest.mark.skipif(sys.platform == "win32",
-                                reason="does not run on windows")
+# pytestmark = pytest.mark.skipif(sys.platform == "win32",
+#                                 reason="does not run on windows")
```
Should this be removed?

johnwparent replied:
Change has been made

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1180,6 +1195,27 @@ def remove_if_dead_link(path):
+
+    implements the callback interface for shutil.rmtree
+
+    Note: This is a no-op on non Windows platforms and will
```
I think this part of the docstring is now out of date (can be removed).

scheibelp replied:
Oh I read more carefully: this *does* say it will throw an error on non-Windows. I think that would mean it isn't a noop though (so in conclusion, can you remove the part that says it's a noop?).

johnwparent replied:
Change made!

scheibelp commented on lib/spack/spack/directory_layout.py:
```diff
@@ -348,6 +350,13 @@ def remove_install_directory(self, spec, deprecated=False):
         """
         path = self.path_for_spec(spec)
         assert(path.startswith(self.root))
+        kwargs = {'ignore_errors': True}
```
Apologies I just realized: unlike `remove_linked_tree`, the invocation of `shutil.rmtree` did not ignore errors originally, so this should probably be more like:

```
if is_windows:
  kwargs = {ignore_errors: false, onerror: fs.error_remove_readonly}
else:
  kwargs= {} # the default value for ignore_errors is false
```

does that seem reasonable?

scheibelp replied:
> Apologies I just realized: unlike remove_linked_tree, the invocation of shutil.rmtree

I meant: "__this__ invocation of shutil.rmtree" i.e. as it's being called in this function

scheibelp commented on lib/spack/spack/directory_layout.py:
```diff
@@ -349,6 +351,15 @@ def remove_install_directory(self, spec, deprecated=False):
 
+        # Windows readonly files cannot be removed by Python
+        # directly. error_remove_readonly respects ignore_error
+        # as True in general while handling the readonly case
```
> respects ignore_error as True

I think that here we generally want to `raise` if the error isn't a permissions error. To that end, `error_remove_readonly` can `raise` if the error isn't the one this PR prepares to handle.

That would mean that `remove_linked_tree` would then generally behave as `ignore_errors=False` on Windows. This would conflict with the test added in https://github.com/spack/spack/pull/29690 (although that test already doesn't run on Windows) but shouldn't cause a regression in the actual fix because that PR also updated how `remove_linked_tree` was called.

Another option is to customize `fs.error_remove_readonly` as a callable object (i.e. make it a `ErrorRemoveReadonly` class with a `__call__` method that either `raise`es or `pass`es depending on how the object was instantiated).

johnwparent replied:
The method has been amended. I liked the callable approach and now a method configured with `ignore_errors` will return a function pointer to the  `error_remove_readonly` method with the `ignore_errors` value set as per the initial call. Let me know if the docstring is unclear.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -1091,15 +1091,35 @@ def visit_directory_tree(root, visitor, rel_path='', depth=0):
+                    # if path is a symlink, determine destination and
+                    # evaluate file vs directory
+                    evald_link = os.readlink(f)
+                    full_path = os.path.join(dir, evald_link)\
```
I don't see the variable `dir` defined elsewhere in this function so I'm surprised that the tests pass.

johnwparent replied:
It's defined on line 1083. `dir = os.path.join(root, rel_path)`

scheibelp replied:
Thanks for the pointer! I was definitely wrong here

billhoffman commented:
@scheibelp  what is it going to take to get this merged? It does cause spack not to work well when you run into it on windows. It is sort of holding up making spack on windows useable IMO. Sorry for jumping in like this ( I was at c++now last week). Anyway would be good to get this merged early this week.

scheibelp commented:
@billhoffman I was communicating privately with @johnwparent as of the end of Friday May 6th about some issues which were resolved this morning.

billhoffman commented:
Great, thanks, glad to see it resolved!
